src/ray/gcs/pubsub/gcs_pub_sub.cc

```diff
@@ -291,5 +295,203 @@ Status PythonGcsPublisher::PublishFunctionKey(
+      if (message.sequence_id() <= max_processed_sequence_id_) {
+        RAY_LOG(WARNING) << "Ignoring out of order message " << message.sequence_id();
+        continue;
+      }
+      max_processed_sequence_id_ = message.sequence_id();
+      if (message.channel_type() != channel_type_) {
+        RAY_LOG(WARNING) << "Ignoring message from unsubscribed channel "
+                         << message.channel_type();
+        continue;
+      }
+      queue_.emplace_back(std::move(message));
+    }
+  }
+
+  *message = queue_.front();
+  queue_.pop_front();
+
+  return Status::OK();
+}
+
+Status PythonGcsSubscriber::PollError(std::string *key_id,
+                                      int64_t timeout_ms,
+                                      rpc::ErrorTableData *data) {
+  rpc::PubMessage message;
+  RAY_RETURN_NOT_OK(DoPoll(timeout_ms, &message));
+  *key_id = message.key_id();
+  *data = message.error_info_message();
+  return Status::OK();
+}
+
+Status PythonGcsSubscriber::PollLogs(std::string *key_id,
+                                     int64_t timeout_ms,
+                                     rpc::LogBatch *data) {
+  rpc::PubMessage message;
+  RAY_RETURN_NOT_OK(DoPoll(timeout_ms, &message));
+  *key_id = message.key_id();
+  *data = message.log_batch_message();
+  return Status::OK();
+}
+
+Status PythonGcsSubscriber::PollFunctionKey(std::string *key_id,
+                                            int64_t timeout_ms,
+                                            rpc::PythonFunction *data) {
+  rpc::PubMessage message;
+  RAY_RETURN_NOT_OK(DoPoll(timeout_ms, &message));
+  *key_id = message.key_id();
+  *data = message.python_function_message();
+  return Status::OK();
+}
+
+Status PythonGcsSubscriber::PollActor(std::string *key_id,
+                                      int64_t timeout_ms,
+                                      rpc::ActorTableData *data) {
+  rpc::PubMessage message;
+  RAY_RETURN_NOT_OK(DoPoll(timeout_ms, &message));
+  *key_id = message.key_id();
+  *data = message.actor_message();
+  return Status::OK();
+}
+
+Status PythonGcsSubscriber::Close() {
+  std::shared_ptr<grpc::ClientContext> current_polling_context;
+  {
+    absl::MutexLock lock(&mu_);
+    closed_ = true;
+    current_polling_context = current_polling_context_;
+  }
+  if (current_polling_context) {
+    current_polling_context->TryCancel();
+  }
+
+  grpc::ClientContext context;
+
+  rpc::GcsSubscriberCommandBatchRequest request;
+  request.set_subscriber_id(subscriber_id_);
+  request.set_sender_id(worker_id_);
+  auto *cmd = request.add_commands();
+  cmd->set_channel_type(channel_type_);
+  cmd->mutable_unsubscribe_message();
+
+  rpc::GcsSubscriberCommandBatchReply reply;
+  grpc::Status status =
+      pubsub_stub_->GcsSubscriberCommandBatch(&context, request, &reply);
+
+  if (!status.ok()) {
+    RAY_LOG(DEBUG) << "Error while closing the subscriber: " << status.error_message()
+                   << " [code " << status.error_code() << "]";
```

Comment:
Should we propagate the error here?

Reply:
Again I'm replicating the behavior of the existing code here.